src/System.Linq.Queryable/src/System/Linq/EnumerableQuery.cs

```diff
@@ -135,7 +135,10 @@ private IEnumerator<T> GetEnumerator()
                 EnumerableRewriter rewriter = new EnumerableRewriter();
                 Expression body = rewriter.Visit(_expression);
                 Expression<Func<IEnumerable<T>>> f = Expression.Lambda<Func<IEnumerable<T>>>(body, (IEnumerable<ParameterExpression>)null);
-                _enumerable = f.Compile()();
+                IEnumerable<T> enumerable = f.Compile()();
+                if (enumerable == this)
+                    throw new InvalidOperationException();
```

Comment:
I'll add a message.
In similar cases,incidentally, we throw a rather cryptic ArgumentException. I'm not sure how much it would break to change those to InvalidOperationException.